profiler/pgo_test.go

```diff
@@ -0,0 +1,120 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2024 Datadog, Inc.
+
+// Only run this test on Go 1.21 and later. This is when pgo became enabled by
+// default in the go compiler.
+//go:build go1.21
+
+package profiler
+
+import (
+	"bytes"
+	"fmt"
+	"go/ast"
+	"go/parser"
+	"go/printer"
+	"go/token"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"runtime/pprof"
+	"strings"
+	"testing"
+	"text/template"
+
+	"github.com/stretchr/testify/require"
+)
+
+// Test_pgoTag verifies that the pgoTag function returns the expected value
+// depending on whether the binary was built with PGO enabled or not.
+func Test_pgoTag(t *testing.T) {
+	var h pgoTestHelper
+	mainSrc := h.mainSource(t)
+	require.Equal(t, "pgo:false", h.goRun(t, mainSrc, nil))
+	require.Equal(t, "pgo:true", h.goRun(t, mainSrc, h.cpuProfile(t)))
+}
+
+// pgoTestHelper is used to group together test helper functions for Test_pgoTag.
+type pgoTestHelper struct{}
+
+// cpuProfile returns a valid CPU profile. No attempt is made to populate it
+// with data.
+func (pgoTestHelper) cpuProfile(t *testing.T) []byte {
+	var buf bytes.Buffer
+	require.NoError(t, pprof.StartCPUProfile(&buf))
+	pprof.StopCPUProfile()
+	data := buf.Bytes()
+	require.NotZero(t, len(data))
+	return data
+}
+
+// mainSource returns the source code for a main package that prints the result
+// of calling pgoTag.
+func (pgoTestHelper) mainSource(t *testing.T) string {
+	// The code below extracts the source of the pgoTag function from the source
+	// of this package and uses it to generate a main package that prints the
+	// result of calling it. This is a bit hacky, but I think it strives a
+	// reasonable balance compared to the alternatives I see:
+	//
+	// a) Not having a test: I considered this, but I want to test this since
+	//    the "-pgo" debug info is not officially documented right now.
+	// b) Full integration test importing dd-trace-go from main.go: This seemed
+	//    like it would be a lot more code and complexity.
```

Comment:
As a third option, we could also put the `pgoEnabled` function in another package and have a test program import it. We'd still need the scaffolding to generate a profile and run the program with/without PGO, but we wouldn't need to pull out big guns to copy/paste functions from this package here. I could try this as a followup if it sounds like a good idea to you.

Reply:
I considered a separate package approach, but thought it would also end up convoluted due to this package probably being `internal`. But maybe it's better? Feel free to send a PR :)